### PR TITLE
Retrieving MongoDB fields blog post.

### PR DESCRIPTION
**Pull request by pauloouriques**
- Add Paulo Ouriques info to authors file;
- Add MongoDB blog post file.

### CODE REVIEW

File: data/authors.ts [hunk 1]
```diff
@@ -62,6 +62,15 @@ const Authors: Author[] = [
     description:
       "Pedro is a full-stack Software Engineer that fell in love with programming at an early age. He's a tinkerer that enjoys reverse engineering, automation and technology in general. When he's not programming, Pedro spends his time playing and listening to music.",
   },
+  {
+    name: "Paulo Ouriques",
+    slug: "pauloouriques",
+    image: "paulo_ouriques.jpg",
+    url: "https://pauloouriques.com/",
```
Comment (evantahler) [inline data/authors.ts:69]: FYI your personal site seems to have an HTTPS problem! 
Screen Shot 2021-04-07 at 9 41 03 AM

File: pages/blog/extracting-all-mongodb-fields.mdx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+In our case, the information to be retrieved is all the field names including the nested ones. Ok, let’s split this problem into two parts: get all first-level fields and get all nested fields with dot notation (i.e. parent.child). For the first one, there are quite a few alternatives to reach the same results, such as:
+
+### Map-Reduce
+
+Map-reduce operations use custom JavaScript functions to map, or associate, values to a key. If a key has multiple values mapped to it, the operation reduces the values for the key to a single object. Here is an example of how to get all first level fields of a collection using Map-Reduce:
+
+```
```
Comment (evantahler) [inline pages/blog/extracting-all-mongodb-fields.mdx]: Use "```js" to get syntax highlighting!

File: pages/blog/extracting-all-mongodb-fields.mdx [hunk 3]
```diff
@@ -0,0 +1,122 @@
+---
+title: "Extracting all MongoBD fields including the nested ones"
+date: "2021-04-06"
+author: "Paulo Ouriques"
+pullQuote: "A simple way to get all first-level fields and, potentially, all nested fields from a MongoDB collection."
+image: "bigquery-source/source-types.png"
+tags: [engineering]
+---
+
+If you’re a data analyst, data scientist, developer, or DB administrator you may have used, at some point, a non-relational database with flexible schemas. Well, I could list several advantages of a NoSQL solution over SQL-based databases. However, the main focus of this post is to discuss a particular downside of MongoDB and a possible solution to go through it.
```
Comment (evantahler) [inline pages/blog/extracting-all-mongodb-fields.mdx]: Ha! I don't think Grouparoo (the company) would say noSQL > SQL.  But, this is Paulo's post, so OK!

File: pages/blog/extracting-all-mongodb-fields.mdx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+Recently, working on the Grouparoo source plugin for MongoDB, I came across a very simple requirement (simple, at least on SQL-based DBs with fixed schemas) which is to get all field names of a collection. However, this is not a simple query on MongoDB as it should be and we’re going to see what is needed to achieve this.
```
Comment (evantahler) [inline pages/blog/extracting-all-mongodb-fields.mdx]: Add a link to the plugin (npm or github, your choice)

Comment (evantahler) [reply pages/blog/extracting-all-mongodb-fields.mdx]: You could add a screenshot showing why you wanted the columns - perhaps from https://github.com/grouparoo/grouparoo/pull/1600

File: pages/blog/extracting-all-mongodb-fields.mdx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+```
+const firstLevelFieldsResult = await db.my_collection.aggregate([
+ { $project: { keyValue: { $objectToArray: "$$ROOT" } } },
+ { $unwind: "$keyValue" },
+ { $group: { _id: null, keys: { $addToSet: "$keyValue.k" } } }
+]);
+```
+
+There are several other ways to achieve the same result, but the aggregation pipeline solution provides better performance (at least than the Map-reduce solution) and usability than the other ones.
```
Comment (evantahler) [inline pages/blog/extracting-all-mongodb-fields.mdx]: Grammar things:

> There are several other ways to achieve the same result, but the aggregation pipeline solution provides better performance (at compared to the map-reduce solution) and (better? consistent?) usability compared to the other options.

File: pages/blog/extracting-all-mongodb-fields.mdx [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+Great!! The first part of our problem is solved. Now, how can we get the nested fields? To solve this, we’re going to get deep into these nested fields and extract all the names one by one. Yup, you’re right, this is a very expensive task and we need to limit the number of documents to get into.
+
+```
+const docs = await db.my_collection.aggregate([
+ { $sample: { size: 10 } },
```
Comment (evantahler) [inline pages/blog/extracting-all-mongodb-fields.mdx]: What's cool about this is that this lets you control the "accuracy" of your search - add more `size` if your schema has dramatically different object types.  I imagine that most of Grouparoo's users are using something Mongose to enforce a schema... so hopefully it's not so crazy.

**Review by evantahler — CHANGES_REQUESTED** [review]
Great start, and good storytelling.  Some minor notes below.

**Comment by evantahler** [comment]
Note that the linter wants you to change something - you can run `prettier --write /path/to/file` to have it fix it up for you (or VS code can be configured to do this whenever you save a file)

File: pages/blog/extracting-all-mongodb-fields.mdx (hunk 3 above)
Comment (pauloouriques) [reply pages/blog/extracting-all-mongodb-fields.mdx]: hahah

Comment (pauloouriques) [reply pages/blog/extracting-all-mongodb-fields.mdx]: I added a vice versa there.

File: pages/blog/extracting-all-mongodb-fields.mdx (hunk 6 above)
Comment (pauloouriques) [reply pages/blog/extracting-all-mongodb-fields.mdx]: Yes, I think the documents will only be slightly different.
But, if we receive any complaints regarding missing fields, we can perform a single-time check (trying to determine how different the docs are) to define the sample size.

**Review by evantahler — APPROVED** [review]

File: pages/blog/extracting-all-mongodb-fields.mdx [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    }
+  },
+  reduce: function (key, stuff) {
+    return null;
+  },
+  out: "my_collection_keys",
+});
+```
+
+After that, run the distinct command on the resulting collection to get the a set containing all keys:
```
Comment (andyjih) [inline pages/blog/extracting-all-mongodb-fields.mdx]:
```suggestion
After that, run the distinct command on the resulting collection to get a set containing all keys:
```

File: pages/blog/extracting-all-mongodb-fields.mdx [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+```js
+const firstLevelFieldsResult = await db.my_collection.aggregate([
+  { $project: { keyValue: { $objectToArray: "$$ROOT" } } },
+  { $unwind: "$keyValue" },
+  { $group: { _id: null, keys: { $addToSet: "$keyValue.k" } } },
+]);
+```
+
+There are several other ways to achieve the same result, but the aggregation pipeline solution provides better performance (at compared to the map-reduce solution) and consistent usability than the other ones.
```
Comment (andyjih) [inline pages/blog/extracting-all-mongodb-fields.mdx]:
```suggestion
There are several other ways to achieve the same result, but the aggregation pipeline solution provides better performance (as compared to the map-reduce solution) and consistent usability than the other ones.
```

**Review by andyjih — APPROVED** [review]
Looks great! Just a few minor tweaks. Thanks Paulo!

**Pull request merged**